sentinel/src/lib/history.js

```diff
@@ -0,0 +1,66 @@
+/**
+ * Utility to track duplicate keys in a certain period of time (periodMins).
+ * Removes expired keys after threshold is reached (purgeThreshold).
+ **/
+const moment = require('moment');
+
+let records = {};
+
+const now = () => {
+  return moment(new Date());
+};
+
+const expired = (key) => {
+  const value = records[key];
+  if(value && value.timestamp) {
+    // Checks difference with history duration in minutes
+    return (now().diff(value.timestamp))/60000 >= module.exports.periodMins;
+  }
+  return true;
+};
+
+const size = () => {
+  return Object.keys(records).length;
+};
+
+const purge = (force=false) => {
+  if(force || size() >= module.exports.purgeThreshold) {
+    for (const key of Object.keys(records)) {
+      if(expired(key)) {
+        delete records[key];
+      }
+    }
+    return true;
+  }
+  return false;
+};
+
+module.exports = {
+
+  periodMins: 30, //Number of minutes to track
+
+  purgeThreshold: 100, //Number of keys to keep in memory
+
+  track: (key) => { //key to track. returns true for duplicates
```

Comment:
I think this should be called `check` or `isDuplicate` or something. The caller doesn't care about how it's tracked, what they're asking is is this key a dupe.